provider/azure/azure.go

```diff
@@ -0,0 +1,212 @@
+	RootURL      string            `json:"rootUrl"`
+	WorkerGroup  string            `json:"workerGroup"`
+	WorkerConfig *cfg.WorkerConfig `json:"workerConfig"`
+}
+
+func (p *AzureProvider) ConfigureRun(state *run.State) error {
+	instanceData, err := p.metadataService.queryInstanceData()
+	if err != nil {
+		return fmt.Errorf("Could not query instance data: %v", err)
+	}
+
+	document, err := p.metadataService.queryAttestedDocument()
+	if err != nil {
+		return fmt.Errorf("Could not query attested document: %v", err)
+	}
+
+	customBytes, err := base64.StdEncoding.DecodeString(instanceData.Compute.CustomData)
+	if err != nil {
+		return fmt.Errorf("Could not read instance customData: %v", err)
+	}
+
+	customData := &CustomData{}
+	err = json.Unmarshal([]byte(customBytes), customData)
+	if err != nil {
+		return fmt.Errorf("Could not parse customData as JSON: %v", err)
+	}
+
+	state.RootURL = customData.RootURL
+	state.WorkerLocation = map[string]string{
+		"cloud":  "azure",
+		"region": instanceData.Compute.Location,
+	}
+
+	wm, err := p.workerManagerClientFactory(state.RootURL, nil)
+	if err != nil {
+		return fmt.Errorf("Could not create worker manager client: %v", err)
+	}
+
+	workerIdentityProofMap := map[string]interface{}{
+		"document": interface{}(document),
+	}
+
+	err = provider.RegisterWorker(
+		state,
+		wm,
+		customData.WorkerPoolId,
+		customData.ProviderId,
+		customData.WorkerGroup,
+		instanceData.Compute.VMID,
+		workerIdentityProofMap)
+	if err != nil {
+		return err
+	}
+
+	providerMetadata := map[string]string{
+		"vm-id":         instanceData.Compute.VMID,
+		"instance-type": instanceData.Compute.VMSize,
+		"region":        instanceData.Compute.Location,
+	}
+
+	if len(instanceData.Network.Interface) == 1 {
+		iface := instanceData.Network.Interface[0]
+		if len(iface.IPV4.IPAddress) == 1 {
+			addr := iface.IPV4.IPAddress[0]
+			providerMetadata["local-ipv4"] = addr.PrivateIPAddress
+			providerMetadata["public-ipv4"] = addr.PublicIPAddress
+		}
+	}
+
+	state.ProviderMetadata = providerMetadata
+
+	state.WorkerConfig = state.WorkerConfig.Merge(customData.WorkerConfig)
+
+	return nil
+}
+
+func (p *AzureProvider) UseCachedRun(run *run.State) error {
+	return nil
+}
+
+func (p *AzureProvider) SetProtocol(proto *protocol.Protocol) {
+	p.proto = proto
+}
+
+func (p *AzureProvider) checkTerminationTime() bool {
+	evts, err := p.metadataService.queryScheduledEvents()
+	if err != nil {
+		log.Printf("While fetching scheduled-events metadata: %v", err)
+		return false
+	}
+
+	// if there are any events, let's consider that a signal we should go away
+	if evts != nil && len(evts.Events) != 0 {
+		log.Println("Azure Metadata Service says a maintenance event is imminent")
+		if p.proto != nil && p.proto.Capable("graceful-termination") {
+			p.proto.Send(protocol.Message{
+				Type: "graceful-termination",
+				Properties: map[string]interface{}{
+					// termination generally doesn't leave time to finish tasks
```

Comment:
this was a bit confusing to me until I thought of the fact that this way the worker can upload logs and such instead of risking waiting till the end. is that why this is? Maybe add a note about that reasoning here. Otherwise it feels like it makes sense to try to finish tasks

Reply:
Yeah, that's the idea: you get an exception / worker-shutdown and immediate retry instead of (if the task doesn't complete in time) exception / claim-expired and a delay before the retry.

Reply:
Oh, I missed the part about "add a note" -- I'll do so in a push to master.